Add support for filtering by publish IP address

dadoonet opened the pull request:
Allocation filtering by IP only works today using the node host address. But in some cases, you might want to filter using the publish/transport address which could be different.

For example, let say your nodes are like this:

```js
// First node
{
 "host" : "es1.acme.com",
 "transport_address" : "inet[/192.168.1.1:9300]",
 "ip" : "127.0.0.1",
 "http_address" : "inet[/192.168.1.1:9200]",
}
```

```js
// Second node
{
 "host" : "es2.acme.com",
 "transport_address" : "inet[/192.168.1.2:9300]",
 "ip" : "127.0.0.1",
 "http_address" : "inet[/192.168.1.2:9200]",
}
```

You can not filter using ip because it will be the same on all hosts.
If we don't want to support it, we should change this PR by some documentation in http://www.elasticsearch.org/guide/en/elasticsearch/reference/current/index-modules-allocation.html#index-modules-allocation to make sure people understand that only node `ip` is supported.

I'm not super familiar with this part of the code. @bleskes? Would you mind have a look and tell me what kind of test I could add and where I can find an example? Also, is it the right way for fixing that :)

s1monw commented:
@dadoonet are you picking this up?

dadoonet commented:
@bleskes Sorry for the loooooong delay it took. I finally applied your comment and added some tests. Let me know what you think about it now.

bleskes commented on core/src/test/java/org/elasticsearch/cluster/node/DiscoveryNodeFiltersTests.java:
```diff
@@ -115,4 +114,56 @@ public void starMatch() {
+                .build();
+        DiscoveryNodeFilters filters = DiscoveryNodeFilters.buildFromSettings(OR, "xxx.", settings);
+
+        DiscoveryNode node = new DiscoveryNode("name1", "id1", DummyTransportAddress.INSTANCE,
```
I think this relies on NetworkUtils#getLocalHostName returning 127.0.0.1, but I don't think it's guaranteed ...

dadoonet replied:
Ah! I was expecting that the Mock was giving this IP. Will try to come with something else. Thanks.

bleskes commented:
@dadoonet I have some concerns about using the transport address as filtering. Maybe we want the publish address instead? the local bind ip may be the same in some VMs - not my expertise and I'm not sure. Maybe @drewr  can comment...

dadoonet commented:
May be we could do both ? Try first publish address and then fallback to transport?

drewr commented:
I need to think about this a bit more, but I think it should always be the `bind_host`. The `publish_host` can introduce more complexity from network topology external to the machine. However, in theory, if the `publish_host` is misconfigured, the node *shouldn't* have been able to join the cluster, but stranger things have happened.

It could be overly cautious, but I would probably make it explicit: `cluster.routing.allocation.include._publish_ip: 99.99.99.99` and support `_bind_ip` as a synonym for `_ip`.

dadoonet commented:
@bleskes @drewr I pushed a new commit which basically changes `DiscoveryNode#address` field from a `TransportAddress` to a `BoundTransportAddress`.

From this object I can access either the `publish_address` or the `bound_address`.

Let me know what you think.

bleskes commented:
@dadoonet thx for the hard work. I'm a bit uncomfortable with changing the address field in the disco node just for this. Given that the current disco node is using the published address, and that's the one we expose in all our reporting API I think it's fine using it that. Here is my suggestion:

- make `_ip` match either the host ip (like it does now) OR the disco node address (published ip). I don't think it's realistic that people will have collisions here.
- add a dedicated selector `_host_ip` which does what we do now.
- add a dedicated selector `_publish_ip` which only matches the address in the disco node.

Makes sense?

dadoonet commented:
@bleskes I updated the PR. Let me know.

bleskes commented on core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodeFilters.java:
```diff
@@ -64,21 +66,55 @@ public static DiscoveryNodeFilters buildFromKeyValue(OpType opType, Map<String,
         this.filters = filters;
     }
 
+    private Boolean match(String[] values, @Nullable String hostIp, @Nullable String publishIp) {
```
can we make this mehtod just do a simple matching - returns true if any of the values matches any of hostIP and publishIP? then we don't need the Boolean and the null return. We then also push the crazy OpType dependent decision to the calling method where it's clearer why we do it.

bleskes commented:
Thx @dadoonet . Left a small suggestion.

dadoonet commented:
@bleskes updated and rebased based on your comment. Let me know

bleskes commented on core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodeFilters.java:
```diff
@@ -64,21 +66,62 @@ public static DiscoveryNodeFilters buildFromKeyValue(OpType opType, Map<String,
         this.filters = filters;
     }
 
+    private boolean match(String[] values, @Nullable String hostIp, @Nullable String publishIp) {
```
call this matchByIP

bleskes commented on core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodeFilters.java:
```diff
@@ -64,21 +66,62 @@ public static DiscoveryNodeFilters buildFromKeyValue(OpType opType, Map<String,
+    private boolean match(String[] values, @Nullable String hostIp, @Nullable String publishIp) {
+        for (String value : values) {
+            boolean matchIp =
+                    (hostIp != null && Regex.simpleMatch(value, hostIp)) ||
```
Regex.simpleMatch already has a null protection, no need to put it here...

bleskes commented on core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodeFilters.java:
```diff
@@ -64,21 +66,62 @@ public static DiscoveryNodeFilters buildFromKeyValue(OpType opType, Map<String,
-                        }
-                    }
+                // We check both the host_ip or the publish_ip
+                String address = null;
```
call this publishAddress?

bleskes commented on core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodeFilters.java:
```diff
@@ -64,21 +66,62 @@ public static DiscoveryNodeFilters buildFromKeyValue(OpType opType, Map<String,
+                boolean match = match(values, node.getHostAddress(), address);
+
+                if (opType == OpType.AND) {
+                    return match;
```
is this correct? When match is true, we need to keep checking for the other filters, no?

bleskes commented on core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodeFilters.java:
```diff
@@ -64,21 +66,62 @@ public static DiscoveryNodeFilters buildFromKeyValue(OpType opType, Map<String,
+                // We check explicitly only the host_ip
+                Boolean match = match(values, node.getHostAddress(), null);
+                if (opType == OpType.AND) {
+                    return match;
```
see comment above...

bleskes commented on core/src/test/java/org/elasticsearch/cluster/node/DiscoveryNodeFiltersTests.java:
```diff
@@ -115,4 +117,100 @@ public void starMatch() {
+                .put("xxx.tag", "A")
+                .put("xxx._publish_ip", "8.8.8.8")
+                .build();
+        DiscoveryNodeFilters filters = DiscoveryNodeFilters.buildFromSettings(AND, "xxx.", settings);
```
maybe add a utility method that randomizes the order of the filters?

bleskes commented:
Thx @dadoonet . Left some more comments.

dadoonet commented:
@bleskes I squashed, rebased and added a new commit to address your latest comments. As I wrote in the commit message 454bd6d, may be I should move `shuffleSettings()` method in `ESTestCase`?

dadoonet commented:
ping @bleskes ? :)

bleskes commented on core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodeFilters.java:
```diff
@@ -64,21 +66,72 @@ public static DiscoveryNodeFilters buildFromKeyValue(OpType opType, Map<String,
+                    address = ((InetSocketTransportAddress) node.address()).address().getAddress().getHostAddress();
+                }
+
+                Boolean match = matchByIP(values, address, null);
```
I think you mean `boolean`?

dadoonet replied:
indeed!

bleskes commented on core/src/test/java/org/elasticsearch/cluster/node/DiscoveryNodeFiltersTests.java:
```diff
@@ -115,4 +122,112 @@ public void starMatch() {
+    private Settings shuffleSettings(Settings source) {
+        Settings.Builder settings = Settings.settingsBuilder();
+        List<String> keys = new ArrayList(source.getAsMap().keySet());
+        Collections.shuffle(keys);
```
we need to pass in the random from the test

dadoonet replied:
Yeah! Makes sense.

bleskes commented on core/src/test/java/org/elasticsearch/cluster/node/DiscoveryNodeFiltersTests.java:
```diff
@@ -115,4 +122,112 @@ public void starMatch() {
+                .build());
+        DiscoveryNodeFilters filters = DiscoveryNodeFilters.buildFromSettings(AND, "xxx.", settings);
+
+        DiscoveryNode node = new DiscoveryNode("", "", "", "192.1.1.54", localAddress, ImmutableMap.of("tag", "A"), null);
```
maybe add a test where tag is "B" and see that it fails?

bleskes commented on core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodeFilters.java:
```diff
@@ -64,21 +66,72 @@ public static DiscoveryNodeFilters buildFromKeyValue(OpType opType, Map<String,
+                }
+            } else if ("_host_ip".equals(attr)) {
+                // We check explicitly only the host_ip
+                Boolean match = matchByIP(values, node.getHostAddress(), null);
```
boolean

bleskes commented:
Left some minor comments. LGTM. Thx @dadoonet . Is the plan to put this in 2.1?

dadoonet commented:
@bleskes I pushed a new commit. Thanks for your comments. Really helpful!
I'm totally +1 to push this in 2.x branch (2.1 version that is)

bleskes commented:
LGTM. Can you mark the PR with the right versions?

dadoonet commented:
@bleskes I had to add a new commit because of forbidden API. Would you mind giving a final look? 
Thanks!

bleskes commented:
Still LGTM

dadoonet commented:
Pushed also in branch 2.x with 38a36ba091c58af163ee870af146ae04a427f5e7

Thanks @bleskes!
